### PR TITLE
feat(keystone): INFRA-446 supercharged knex DB adapter 🔌💪⛁⛁

### PR DESCRIPTION
**Pull request by SavelevMatthew**
## TODO:

- [x] Write utils/rules tests
- [x] add readme?
- [x] Remove debug schemas in `test` domain
- [x] Enable replication in CI/CD

### CODE REVIEW

File: packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/utils/env.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+const validateDBConfig = ajv.compile(DB_URL_SCHEMA)
+
+/**
+ * Converts custom database url string to dictionary of form Record<db_name, connection_string>
+ *
+ * NOTE: DB_NAME is important here, since it all dbs will be grouped together by names later.
+ * Flat connection string is good, but it can contain query parameters for configuration,
+ * and required to be the same across all others env values, so assigning a name to it is easier to maintain
+ *
+ * @param {string | undefined} databaseUrl - custom db url. Example: 'custom:{"write": "postgresql://postgres:postgres@127.0.0.1:5433/local-condo", "async_replica": "postgresql://postgres:postgres@127.0.0.1:5432/local-condo"}'
```
[the remark this answers is not in the document]
Comment (SavelevMatthew) [reply packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/utils/env.js]: Test local creds, used in JSDoc. Not real credentials. Replaced with ***, so this rule not triggered anymore

[the remark this answers is not in the document]
Comment (SavelevMatthew) [reply packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/utils/env.js]: Test local creds, used in JSDoc. Not real credentials. Replaced with ***, so this rule not triggered anymore

File: bin/prepare.js [hunk 2]
```diff
@@ -117,9 +117,21 @@ async function prepare () {
...
+                })}`
+                env.DATABASE_POOLS = JSON.stringify({
+                    main: { databases: ['main'], writable: true },
+                    replicas: { databases: ['replica'], writable: false },
+                })
+                env.DATABASE_ROUTING_RULES = JSON.stringify([
+                    { target: 'main', gqlOperationType: 'mutation' },
+                    { target: 'replicas', sqlOperationName: 'select' },
+                    { target: 'replicas', sqlOperationName: 'show' },
+                    { target: 'main' },
```
Comment (sitozzz) [inline bin/prepare.js:133]: Last one is for default route, right?

File: packages/keystone/KSv5v6/v5/keystone.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+const { Keystone: DefaultKeystone } = require('@keystonejs/keystone')
+
+const { _patchResolvers } = require('../utils/resolvers')
+
+class Keystone extends DefaultKeystone {
+    getResolvers ({ schemaName }) {
+        const originalResolvers = super.getResolvers({ schemaName })
+
+        return _patchResolvers(originalResolvers)
```
Comment (sitozzz) [inline packages/keystone/KSv5v6/v5/keystone.js:9]: `_patchResolvers` is too general a name for this function. Maybe specify what really you patch here?

File: bin/prepare.js (hunk 2 above)
Comment (SavelevMatthew) [reply bin/prepare.js:133]: Yep, rules is resolved one-by-one

File: packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    async _initKnexClients () {
+        const dbNames = Object.keys(this._dbConnections)
+        const connectionResults = await Promise.allSettled(
+            dbNames.map(dbName => initKnexClient({
+                client: 'postgres',
+                pool: { min: 0, max: 3 },
+                connection: this._dbConnections[dbName],
+            }))
+        )
```
Comment (sitozzz) [inline packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js:37]: Any validations before init knex adapter?

Comment (SavelevMatthew) [reply packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js:37]: All configs are validated inside constructor...
Connection is validated inside `initKnexClient`

What else you expect to validate?

File: packages/keystone/KSv5v6/v5/keystone.js (hunk 3 above)
Comment (SavelevMatthew) [reply packages/keystone/KSv5v6/v5/keystone.js:9]: I'll think about naming here

File: packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js (hunk 4 above)
Comment (sitozzz) [reply packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js:37]: Ok, I found it. Seems like that's enough

File: packages/keystone/package.json [hunk 5]
```diff
@@ -84,6 +86,7 @@
   },
   "repository": "https://github.com/open-condo-software/condo/tree/main/packages/keystone",
   "devDependencies": {
+    "@faker-js/faker": "^7",
```
Comment (sitozzz) [inline packages/keystone/package.json:89]: Is that version compatible with the installed one over the apps? Maybe we should sync versions

**Review by sitozzz — APPROVED** [review]

File: packages/keystone/package.json (hunk 5 above)
Comment (SavelevMatthew) [reply packages/keystone/package.json:89]: Yep, we have 7th on global package.json

**Review by Alllex202 — APPROVED** [review]

File: packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/utils/env.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+const SIMPLE_NAME_PATTERN = /^[a-z\d_+]+$/i
+
+const DB_URL_SCHEMA = {
+    type: 'object',
+    minProperties: 1,
+    additionalProperties: false,
+    patternProperties: {
+        [ANY_CHAR_PATTERNS]: {
+            type: 'string',
+            pattern: '^postgresql://.+',
```
[the remark this answers is not in the document]
Comment (SavelevMatthew) [reply packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/utils/env.js:21]: I'm not sure its necessary right now

File: packages/keystone/KSv5v6/v5/keystone.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (SavelevMatthew) [reply packages/keystone/KSv5v6/v5/keystone.js:9]: I don't know. It depends on how we see future patches. Like...

Option 1 (chain of patches)
```
const originalResolvers = super.getResolvers({ schemaName })
const withGQLContext = _patchResolversWithGQLContext(originalResolvers)
const withSomethingElse = _patchResolversWithSomethingElse(withGQLContext)

return withSomethingElse
```

Option 2 (single patch)
```
const originalResolvers = super.getResolvers({ schemaName })

return _patchResolvers(originalResolvers)
```

[the remark this answers is not in the document]
Comment (SavelevMatthew) [reply packages/keystone/KSv5v6/v5/keystone.js]: Renamed util

**Pull request merged**